sound/soc/sof/topology.c

```diff
@@ -2017,6 +2042,60 @@ static int sof_get_control_data(struct snd_soc_component *scomp,
 	return 0;
 }
 
+/*
+ * Sof mux/demux component
+ */
+static int sof_process_load_mux(struct snd_soc_component *scomp,
+				struct sof_ipc_comp_process *process,
+				struct snd_soc_tplg_private *private)
+{
+	struct sof_mux_config *mc = (struct sof_mux_config *)&process->data;
+	struct mux_stream_data *msd = (struct mux_stream_data *)
+		((u8 *)process->data + sizeof(struct sof_mux_config));
+	struct snd_soc_tplg_vendor_array *varray;
+	int pos = 0;
+	int ret;
+	int i;
+
+	/* parse mux config array */
+	ret = sof_parse_tokens(scomp, mc, mux_tokens,
+			       ARRAY_SIZE(mux_tokens), private->array,
+			       le32_to_cpu(private->size));
+	if (ret < 0) {
+		dev_err(scomp->dev, "error: parse mux tokens failed %d\n",
+			private->size);
+		return ret;
+	}
+
+	/* parse mux stream config arrays */
+	for (i = 0; i < mc->num_streams; i++) {
+		varray = (struct snd_soc_tplg_vendor_array *)
+			((u8 *)private->array  + pos);
+		ret = sof_parse_tokens(scomp, msd, mux_stream_tokens,
+				       ARRAY_SIZE(mux_stream_tokens),
+				       varray,
+				       le32_to_cpu(private->size) - pos);
+		if (ret < 0) {
+			dev_err(scomp->dev, "error: mux stream tokens %d\n",
+				private->size);
+			return ret;
+		}
+
+		/* track position for multiple same type arrays */
```

Comment:
I have idea what 'multiple same type arrays' are? 

As opposed to single same type array? or multiple different type arrays?

Reply:
If you have 2 similar tuples arrays in one component's topology, current parser will always parse the last one. So it will parse the 1st one, but continue and parse the 2nd on top of the 1st. Mux is an example of this kind of component, because it may contain multiple stream config arrays. 

I'm not sure if we currently any other component with this kind of structure, but we may in the future. Also I'm not aware any other way how this should be handled in alsa topology, is it forbidden to define multiple similar tuples arrays for one component? At least currently the alsa tools are fine with it.

Reply:
can you give an example of what 'similar tuple arrays in one components's topology' mean?

These are really vague concepts for me, and I am not sure why this is exposed by the mux/demux support?

Reply:
@plbossart @juimonen If I understand it correctly, I think the problem is similar to what we've done for the PDM controller config in the topology for DMIC config already. We're looking at something like below where we have multiple sets of tokens.

https://github.com/thesofproject/sof/pull/2345#issuecomment-581502285